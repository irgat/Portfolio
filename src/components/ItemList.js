import React from 'react';
import PropTypes from 'prop-types';
import Item from './Item';

const propTypes = {
    items: PropTypes.array.isRequired,
    selectedItem: PropTypes.any.isRequired,
    onClick: PropTypes.func.isRequired,
};

const defaultProps = {

};

class ItemList extends React.Component {
    filterValidItems = (items) => {
        return items.filter(item => (item.thumbnail !== undefined && item.name !== undefined));
    };

    render() {
        let validItems = this.filterValidItems(this.props.items);
        return (
            (validItems.length > 0) && <div>
                {validItems.map((item, index) => (
                    <Item
                        key={index}
                        id={item.name}
                        imagePath={item.thumbnail}
                        isSelected={item.name === this.props.selectedItem}
                        onClick={this.props.onClick}
                        onMouseOver={(item.name !== this.props.selectedItem) ? this.onMouseOver : null}
                        onMouseOut={(item.name !== this.props.selectedItem) ? this.onMouseOut : null}
                    />
                ))}
            </div>
        );
    }
}

ItemList.propTypes = propTypes;
ItemList.defaultProps = defaultProps;

export default ItemList;