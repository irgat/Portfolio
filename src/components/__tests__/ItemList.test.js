import React from 'react';
import Enzyme from 'enzyme';
import ItemList from '../ItemList';

const invalidList = [
    {},
    { key: 'value' },
    { thumbnail: 'folderName/fileName.ext' },
    { name: '#' },
];

const mixedList = [
    { thumbnail: 'folderName/fileName.ext', name: '#' },
    { thumbnail: 'folderName/fileName.ext', name: '##', key: 'value' },
].concat(invalidList);

const itemList = (items) => {
    return Enzyme.shallow(
        <ItemList
            items={items}
            selectedItem='##'
            onClick={jest.fn()}
        />
    );
};

describe('<ItemList />', () => {
    it('filterValidItems() returns 2 items', () => {
        expect(itemList(mixedList).instance().filterValidItems(mixedList)).toMatchSnapshot();
    });

    it('returns null when there is no valid items', () => {
        expect(itemList(invalidList).find('div').getElements()).toMatchSnapshot();
    });

    describe('<div />', () => {
        const wrapper = itemList(mixedList);
        const validItems = wrapper.instance().filterValidItems(mixedList);
        it(`receives ${mixedList.length} item(s), renders ${validItems.length} valid <Item />(s)`, () => {
            expect(wrapper.find('Item')).toHaveLength(validItems.length);
            expect(wrapper.find('div').getElements()).toMatchSnapshot();
        });
    });
});
