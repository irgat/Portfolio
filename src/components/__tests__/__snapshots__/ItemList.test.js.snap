// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> <div /> receives 6 item(s), renders 2 valid <Item />(s) 1`] = `
Array [
  <div>
    <Item
      id="#"
      imagePath="folderName/fileName.ext"
      isSelected={false}
      onClick={[MockFunction]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    />
    <Item
      id="##"
      imagePath="folderName/fileName.ext"
      isSelected={true}
      onClick={[MockFunction]}
      onMouseOut={null}
      onMouseOver={null}
    />
  </div>,
]
`;

exports[`<ItemList /> filterValidItems() returns 2 items 1`] = `
Array [
  Object {
    "name": "#",
    "thumbnail": "folderName/fileName.ext",
  },
  Object {
    "key": "value",
    "name": "##",
    "thumbnail": "folderName/fileName.ext",
  },
]
`;

exports[`<ItemList /> returns null when there is no valid items 1`] = `Array []`;
